caching/memcache.go

```
@@ -99,6 +103,56 @@ func (m *Memcache) Get(k interface{}, target interface{}, fn RetrieveFunc) error
 	return nil
 }
 
+// lockOrSkip is a variation on the lock package's WaitAndAcquire method where each time we try to
+// acquire the lock, we first see if
+func lockOrSkip(ctx context.Context, key string, lockDuration, timeout time.Duration, skip func() bool) (bool, error) {
+	acquire := func() bool {
+		if err := lock.For(ctx, key, lockDuration); err == nil {
+			return true
+		} else if err != lock.ErrAlreadyLocked {
+			log.Errorf(ctx, "lock: error acquiring (%T): %v", err, err)
+		}
+
+		return false
+	}
+
+	if acquire() {
+		return true, nil
+	}
+
+	expr := time.NewTimer(timeout)
+	defer expr.Stop()
+	done := expr.C
+
+	ticker := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				// random sleep of 50-100ms
+				// TODO: make this configurable, add optional backoff functionality
+				time.Sleep((time.Duration)(50+rand.Int63n(50)) * time.Millisecond)
+				ticker <- struct{}{}
+			}
+		}
+	}()
+
+	for {
+		select {
+		case <-done:
+			return false, lock.ErrTimeout
+		case <-ticker:
+			if skip() {
+				return false, nil
+			} else if acquire() {
+				return true, nil
+			}
+		}
+	}
+}
```

Seeing how significant the duplication is now, the idea you suggested offline about adding functionality to the lock package might be a better way to go. Perhaps using the word "abort" rather than "skip" though?